criu/net.c

```diff
@@ -1269,6 +1275,110 @@ static int restore_one_link(struct ns_id *ns, struct net_link *link, int nlsk,
 	return do_rtm_link_req(RTM_NEWLINK, link, nlsk, ns, link_info, extras);
 }
 
+struct move_req {
+	struct newlink_req req;
+	char ifnam[IFNAMSIZ];
+};
+
+static int move_veth_cb(void *arg, int fd, pid_t pid)
+{
+	int fd_ns_old = -1, ret = -1;
+	struct move_req *mvreq = arg;
+	struct newlink_req *req = &mvreq->req;
+	int ifindex, nlsk;
+
+	if (!(root_ns_mask & CLONE_NEWUSER)) {
+		int fd_ns;
+
+		fd_ns = get_service_fd(NS_FD_OFF);
+		if (switch_ns_by_fd(fd_ns, &net_ns_desc, &fd_ns_old))
+			return -1;
+	}
+
+	/* Retrieve ifindex of precreated veth device in source netns. */
+	ifindex = if_nametoindex(mvreq->ifnam);
+	if (!ifindex)
+		goto out;
+	req->i.ifi_index = ifindex;
+
+	/* Tell netlink what netns we want to move that veth device into. */
+	addattr_l(&req->h, sizeof(*req), IFLA_NET_NS_FD, &fd, sizeof(fd));
+
+	nlsk = socket(PF_NETLINK, SOCK_RAW | SOCK_CLOEXEC, NETLINK_ROUTE);
+	if (nlsk < 0)
+		goto out;
+
+	ret = do_rtnl_req(nlsk, req, req->h.nlmsg_len, restore_newlink_cb, NULL, NULL, NULL);
+	close(nlsk);
+
+out:
+	if (fd_ns_old >= 0)
+		ret = restore_ns(fd_ns_old, &net_ns_desc);
+
+	return ret;
+}
+
+static int move_veth(const char *netdev, struct ns_id *ns,
+		     struct net_link *link, int nlsk)
+{
+
+	NetDeviceEntry *nde = link->nde;
+	struct newlink_req *req;
+	struct move_req mvreq;
+	size_t len_val;
+	int ret;
+
+	/*
+	 * We require a target ifindex otherwise we can't restore addresses
+	 * later on as ip stores ifindex in its address dump for network
+	 * devices.
+	 */
+	if (!nde->ifindex)
+		return -1;
+
+	memset(&mvreq.req, 0, sizeof(mvreq.req));
+	req = &mvreq.req;
+
+	req->h.nlmsg_len	= NLMSG_LENGTH(sizeof(struct ifinfomsg));
+	req->h.nlmsg_flags	= NLM_F_REQUEST|NLM_F_ACK;
+	req->h.nlmsg_type	= RTM_NEWLINK;
+	req->h.nlmsg_seq	= CR_NLMSG_SEQ;
+
+	req->i.ifi_family	= AF_UNSPEC;
+	req->i.ifi_flags	= nde->flags;
+
+	/* Tell netlink what name we want in the target netns. */
+	addattr_l(&req->h, sizeof(*req), IFLA_IFNAME,
+		  nde->name, strlen(nde->name));
+
+	/* Tell netlink what mtu we want in the target netns. */
+	addattr_l(&req->h, sizeof(*req), IFLA_MTU,
+		  &nde->mtu, sizeof(nde->mtu));
+
+	/* Tell netlink what ifindex we want in the target netns. */
+	addattr_l(&req->h, sizeof(*req), IFLA_NEW_IFINDEX,
```

Comment:
I'm not sure we need feature checking for this. If this property isn't supported on the kernel then we'll simply fail the restore which I think is reasonable.

Reply:
as for me, we can just write a notice about that in option description